ethers-contract/src/error.rs

```diff
@@ -1,11 +1,34 @@
 use ethers_core::{
     abi::{AbiDecode, AbiEncode, Tokenizable},
-    types::{Bytes, Selector},
+    types::Selector,
     utils::id,
 };
 use ethers_providers::JsonRpcError;
 use std::borrow::Cow;
 
+/// A trait for enums unifying [`EthError`] types.
+///
+/// We do not recommend manual implementations of this trait. Instead, use the
+/// automatically generated implementation in the [`crate::abigen`] macro
+pub trait ContractRevert: AbiDecode + AbiEncode + Send + Sync {
+    /// Decode the error from EVM revert data including an Error selector
```

Comment:
is this only derived for enums?
because this clashes with `EthError`

Reply:
yes, this trait is only for unifying multiple potential ethereum errors into 1 type. Basically this should be thought of as "the errors associated with a specific contract's ABI". 

I thought about breaking out a super-trait that had the shared methods on it, and decided that it was too complex for users

Reply:
i also thought about modifying AbiDecode to include `decode_with_selector(selector: Selector, bytes: &[u8])` but didn't want to get into lower level changes

Reply:
you can also consult the docstring, btw

```
/// A trait for enums unifying [`EthError`] types.
///
/// We do not recommend manual implementations of this trait. Instead, use the
/// automatically generated implementation in the [`crate::abigen`] macro```